backend/copr_backend/pulp.py

```diff
@@ -10,6 +10,148 @@
+        plans to finish the task right away).
+        """
+        if self.noop:
+            return False
+
+        status = self.redis.hget(self.key, "status") == "success"
+        self.log.debug("Has already a status? %s", status)
+
+        try:
+            if not status:
+                # not yet processed
+                return False
+        finally:
+            # This is atomic operation, other processes may not re-start doing this
+            # task again.  https://github.com/redis/redis/issues/9531
+            #
+            # The 'delete_if_not = True' is not a TOCTOU case.  Our caller holds
+            # the lock, hence no other machine may take our task (would require
+            # lock) and re-process it.
+            if status or delete_if_not:
+                self.redis.delete(self.key)
+
+        return status
+
+    def options(self):
+        """
+        Get the options from other _compatible_ (see below) Redis tasks, and
+        plan the list of tasks in self.notify_keys[] that we will notify in
+        commit().
+        """
+        rpms_to_add = set(self.rpms_to_add)
+        rpms_to_remove = set(self.rpms_to_remove)
+
+        if self.noop:
+            return {"add_content_units": rpms_to_add, "remove_content_units": rpms_to_remove}
+
+        for key in self.redis.keys(self.key_pattern):
+            assert key != self.key
+
+            task_dict = self.redis.hgetall(key)
+            if not task_dict:
+                # TOCTOU: The key might no longer exist in DB, even though
+                # _our process_ holds the lock!  See the
+                # 'self.redis.delete(self.key)' above in check_processed(); the
+                # _original process_ removes key without locking, if _third_
+                # process commit()ted the status=succeeded.
+                # Prior fixing #3770/#3777 we tried to process these deleted
+                # tasks and ended up with KeyError on task_dict["task"] below.
+                self.log.info("Key %s already processed (by other process), "
+                              "removed by the original process (without lock) "
+                              "since we listed it.", key)
+                continue
+
+            if task_dict.get("status") is not None:
+                # skip processed tasks
+                self.log.info("Key %s already processed, skip", key)
+                continue
+
+            task_opts = json.loads(task_dict["task"])
+
+            # we can process this task!
+            self.notify_keys.append(key)
+
+            # append "add" tasks, if that makes sense
+            rpms_to_add.update(task_opts["add_content_units"])
+            rpms_to_remove.update(task_opts["remove_content_units"])
+
+            if len(self.notify_keys) >= MAX_IN_BATCH - 1:  # one is ours!
+                self.log.info("Batch add-remove limit %s reached, skip the rest",
+                              MAX_IN_BATCH)
+                break
```

Comment:
I think we can drop this "MAX" thing.  The reason we do that for backend storage is that we use subprocess & execve() that has some commandline limits, but I doubt we have limits for json API payload size?

The thing is, especially for removals, that you get a very large json here... consider user selects thousands of builds for deletion (single button), each build has many chroot-builds, and each chroot build many prns.

Reply:
i don't remember what the maximum size for a POST request payload is, but we should probably not worry about that right now. if we run into it at some point we can add a max limit here.